### PR TITLE
feat: Display better error messages (STARGATE-1953)

### PR DESCRIPTION
**Pull request by lukewhyte**
This is ready for review but I'm leaving it as a a draft because I want to confirm a couple things with backend team and I have a question about if or where we should be catching these errors.

### CODE REVIEW

File: packages/pci-sdk-iframe/src/apiClient/index.ts [hunk 1]
```diff
@@ -93,11 +100,16 @@ export async function getCardData(
 		headers,
 		body,
 	})
-		.then((res) => {
-			// TODO: Handle different errros better
```
Comment (lukewhyte) [inline packages/pci-sdk-iframe/src/apiClient/index.ts:97]: I always liked that the TODO to handle errors better had an error in the word "error". I'm sad to see this one go.

File: packages/pci-sdk-iframe/src/apiClient/errors.service.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+interface I401Error {
+	message: string;
+	code: number;
+}
+
+async function get401ResponseMessage(res: Response): Promise<string> {
```
Comment (IagoLast) [inline packages/pci-sdk-iframe/src/apiClient/errors.service.ts]: 🔴 We are leaking an implementation detail into the message parser.

IMO the message parser should receive a number a return a string / Error

File: packages/pci-sdk-iframe/src/apiClient/index.ts [hunk 3]
```diff
@@ -23,12 +24,14 @@ export async function request2FACode(): Promise<IRequest2FACodeResponse> {
 		headers,
 		body: JSON.stringify({ show_verification_secret: true }),
 	})
-		.then((res) => {
+		.then(async (res) => {
```
Comment (IagoLast) [inline packages/pci-sdk-iframe/src/apiClient/index.ts]: I think the easiest method would be:

```ts
case: 401:
  const {code} = await res.json();
  const errorMessage = errorsService. get401ResponseMessage(code);
  throw new Error(errorMessage);
```

Comment (IagoLast) [reply packages/pci-sdk-iframe/src/apiClient/index.ts]: also we could refactor this a little bit and parse the json just one time

File: packages/pci-sdk-iframe/src/apiClient/index.ts [hunk 4]
```diff
@@ -53,12 +56,16 @@ export async function verify2FACode(secret: string, verificationId: string): Pro
...
 			switch (res.status) {
 				case 200:
 					return res.json();
+				case 401:
+					throw new Error(await errorsService.get401ResponseMessage(res));
+				case 400:
+					throw new Error('Invalid request. Please contact APTO Payments');
 				default:
-					throw Error(res.status.toString());
+					throw new Error(res.status.toString());
```
Comment (IagoLast) [inline packages/pci-sdk-iframe/src/apiClient/index.ts]: Since this code is duplicated we could extract a function here

**Review by IagoLast — CHANGES_REQUESTED** [review]
- If you need a code don't inject a response (🐒 🍌 )
- The error handler code is duplicated, we could extract it to a function
- I think we could refactor the apiClient code a little bit using async instead promises

File: packages/pci-sdk-iframe/src/apiClient/errors.service.ts (hunk 2 above)
Comment (lukewhyte) [reply packages/pci-sdk-iframe/src/apiClient/errors.service.ts]: > I think we could refactor the apiClient code a little bit using async instead promises

@IagoLast I think this is the solution. Eslint was not being my friend during this process, which is why I ended up leaking the whole response into the parser.

File: packages/pci-sdk-iframe/src/apiClient/index.ts (hunk 3 above)
Comment (lukewhyte) [reply packages/pci-sdk-iframe/src/apiClient/index.ts]: @IagoLast This was exactly what I tried at first but Eslint won't let you put const, let or var in a switch block.

**Review by IagoLast — APPROVED** [review]

**Pull request merged**